### PR TITLE
samples: pmic: npm6001: Add nPM6001 PMIC sample

### PR DESCRIPTION
**Pull request by nordic-auko**
Add nPM6001 PMIC sample containing a sample nPM6001 driver,
which is designed to be portable to non-NCS platforms.

Sample requires nPM6001-EK connected to a nRF5x-DK,
and demonstrates TWI communication and nPM6001 features.

Signed-off-by: Audun Korneliussen <audun.korneliussen@nordicsemi.no>

### CODE REVIEW

**Comment by nordic-auko** [comment]
It was decided that the nPM6001 sample should be set up as more of a standalone sample instead of splitting it up as sample + zephyr/ncs driver. This way the pmic reference code is more accessible outside of NCS.

npm6001.h and npm6001_types.h are autogenerated files. At some point it might be worthwhile to have these in nrfx/mdk (alongside the nRF series register map files) instead of in the sample.

**Review by shantha-14 — CHANGES_REQUESTED** [review]
Downstream execution on HW target missing

**Comment by nordic-auko** [comment]
> Downstream execution on HW target missing

I added a loopback test case under console harness in sample.yaml for now. Once we get availability of npm6001 devkit hardware I can work on setting up a harness for this on test servers in TRD

File: samples/pmic/npm6001/src/sample_shell_cmds.c [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(npm6001_cmds,
+	SHELL_CMD_ARG(vreg, &npm6001_vreg_cmds, "Voltage regulator control", cmd_vreg, 2, 1),
+	SHELL_CMD_ARG(reg, &npm6001_reg_cmds, "Register read/write", NULL, 1, 2),
+	SHELL_CMD_ARG(
+		watchdog, &npm6001_wd_cmds, "Watchdog enable/disable/kick", cmd_wd_enable, 1, 1),
+	SHELL_CMD_ARG(hibernate, NULL, "Hibernate", cmd_hibernate, 1, 1),
+	SHELL_CMD_ARG(interrupt, &npm6001_int_cmds, "Interrupt enable/disable", cmd_int, 2, 1),
+
```
Comment (erikrobstad) [inline samples/pmic/npm6001/src/sample_shell_cmds.c:534]: Remove newline

**Review by erikrobstad — CHANGES_REQUESTED** [review]

File: samples/pmic/npm6001/README.rst [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+   * - Switch
+     - Position
+   * - SW9 (nRF power source)
+     - VDD
+   * - SW10 (VEXT -> nRF)
+     - ON
+   * - SW6
+     - DEFAULT
+
+.. figure:: /images/pmic_npm6001ek_nrf53840dk.svg
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/README.rst:113]: Yeah, there are some changes in the hardware, with a new minor update on the way. Will update the figure to match the final hardware.

File: samples/pmic/npm6001/sample.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+sample:
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/sample.yaml:1]: Yes, good point. Adding this.

File: doc/nrf/samples/samples_pmic.rst [hunk 4]
```diff
@@ -0,0 +1,11 @@
+.. _pmic_samples:
+
+PMIC samples
+################
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply doc/nrf/samples/samples_pmic.rst:4]: Fixed

File: samples/pmic/npm6001/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,28 @@
+#
+# Copyright (c) 2022 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+cmake_minimum_required(VERSION 3.20.0)
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/CMakeLists.txt:6]: Fixed

File: samples/pmic/npm6001/src/main.c [hunk 6]
```diff
@@ -0,0 +1,506 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/main.c:5]: Fixed

File: samples/pmic/npm6001/src/main.c [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/logging/log.h>
+
+#include <dk_buttons_and_leds.h>
+
+#include "drv_npm6001.h"
+
+LOG_MODULE_REGISTER(app, CONFIG_APP_LOG_LEVEL);
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/main.c:16]: Changed to main

File: samples/pmic/npm6001/src/main.c [hunk 8]
```diff
@@ -0,0 +1,506 @@
...
+	err = i2c_transfer(i2c_dev, &msgs[0], ARRAY_SIZE(msgs), DRV_NPM6001_TWI_ADDR);
+	if (err == 0) {
+		/* Success: return number of bytes written */
+		return len;
+	} else {
+		return -1;
+	}
+}
+
+/**@brief Print sample help message to log
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/main.c:99]: Changing as suggested

File: samples/pmic/npm6001/src/main.c [hunk 9]
```diff
@@ -0,0 +1,506 @@
...
+				LOG_ERR("drv_npm6001_genio_set=%d", err);
+			} else {
+				genio0_set = true;
+				(void) dk_set_led_on(LED_GENIO0_ON);
+			}
+		}
+	}
+
+	if (button_state == DK_ALL_BTNS_MSK) {
+		/* Hibernate for 8 seconds (= 2 ticks) */
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/main.c:197]: Yes, both the button function and the duration

File: samples/pmic/npm6001/src/npm6001/drv_npm6001.c [hunk 10]
```diff
@@ -0,0 +1,1177 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include "drv_npm6001.h"
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/npm6001/drv_npm6001.c:6]: Fixed

File: samples/pmic/npm6001/src/npm6001/drv_npm6001.c [hunk 11]
```diff
@@ -0,0 +1,1177 @@
...
+
+#include <errno.h>
+
+#ifndef ROUNDED_DIV
+#define ROUNDED_DIV(a, b)  (((a) + ((b) / 2)) / (b))
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/npm6001/drv_npm6001.c:11]: Fixed

File: samples/pmic/npm6001/src/npm6001/npm6001_types.h [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+#define __IM volatile const
+#endif
+#ifndef __OM
+#define __OM volatile
+#endif
+#ifndef __IOM
+#define __IOM volatile
+#endif
+
+#include "compiler_abstraction.h"
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/npm6001/npm6001_types.h:41]: This file is auto-generated. Will ping the authors

File: samples/pmic/npm6001/src/npm6001/drv_npm6001.c [hunk 13]
```diff
@@ -0,0 +1,1177 @@
...
+
+	bytes_written = twi_write(&reg_val, sizeof(reg_val), reg_addr);
+	if (bytes_written != sizeof(reg_val)) {
+		return -ENODEV;
+	} else {
+		return 0;
+	}
+
+	/* Revert PWM mode if needed */
+	if (pwm_mode_enable) {
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/npm6001/drv_npm6001.c:182]: Good catch

File: samples/pmic/npm6001/src/npm6001/drv_npm6001.c [hunk 14]
```diff
@@ -0,0 +1,1177 @@
...
+			VOLTAGE_VALID(voltage, DRV_NPM6001_DCDC3_MINV, DRV_NPM6001_DCDC3_MAXV);
+		voltage_res = DRV_NPM6001_DCDC3_RES;
+		break;
+	case DRV_NPM6001_LDO0:
+		voltage_valid =
+			VOLTAGE_VALID(voltage, DRV_NPM6001_LDO0_MINV, DRV_NPM6001_LDO0_MAXV);
+		voltage_res = DRV_NPM6001_LDO0_RES;
+		break;
+	case DRV_NPM6001_LDO1:
+		/* LDO1 cannot be adjusted */
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/npm6001/drv_npm6001.c:556]: I intended to have fall-through to the default case and return error. Adding comment

File: samples/pmic/npm6001/src/npm6001/drv_npm6001.c [hunk 15]
```diff
@@ -0,0 +1,1177 @@
...
+		reg_val = DIGITAL_LDO0VOUT_VOLTAGE_SET3V3 << DIGITAL_LDO0VOUT_VOLTAGE_Pos;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	reg_addr = NPM6001_REG_ADDR(LDO0VOUT);
+
+	bytes_written = twi_write(&reg_val, sizeof(reg_val), reg_addr);
+	if (bytes_written != sizeof(reg_val)) {
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/npm6001/drv_npm6001.c:223]: Good point. Changing this here and other places.

File: samples/pmic/npm6001/src/npm6001/drv_npm6001.c [hunk 16]
```diff
@@ -0,0 +1,1177 @@
...
+	uint8_t reg_val;
+	uint8_t reg_addr;
+	int bytes_read;
+
+	if (timer_val == 0 || timer_val > 0x00FFFFFF) {
+		return -EINVAL;
+	}
+
+	/* Check if watchdog is running.
+	 * If yes, it must be disabled to ensure hibernation duration is set properly.
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/npm6001/drv_npm6001.c:1158]: No, I believe everything is at reset values after hibernation. Will check this

File: samples/pmic/npm6001/src/npm6001/drv_npm6001.h [hunk 17]
```diff
@@ -0,0 +1,449 @@
...
+		DRV_NPM6001_GENIO_CFG_SENSE_LOW,
+		DRV_NPM6001_GENIO_CFG_SENSE_HIGH,
+	} sense;
+};
+
+/**
+ * @brief Platform-specific functions for TWI and GPIO control.
+ */
+struct drv_npm6001_platform {
+	/**@brief Initialize platform-specifics. Can be NULL if not needed.
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/npm6001/drv_npm6001.h:200]: Also for "one-liners" such as `/**@brief nPM6001 TWI address. */`?

File: samples/pmic/npm6001/src/npm6001/drv_npm6001.h [hunk 18]
```diff
@@ -0,0 +1,449 @@
...
+ * @return 0 if successful. Otherwise a negative error code.
+ */
+int drv_npm6001_vreg_mode_pin_enable(
+	enum drv_npm6001_vreg regulator,
+	enum drv_npm6001_mode_pin pin);
+
+/**@brief Configure mode pin.
+ *
+ * @param[in] pin Mode pin.
+ *
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/npm6001/drv_npm6001.h:305]: Fixed

File: samples/pmic/npm6001/src/sample_shell_cmds.c [hunk 19]
```diff
@@ -0,0 +1,546 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include <stdlib.h>
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/sample_shell_cmds.c:6]: Fixed

File: samples/pmic/npm6001/src/sample_shell_cmds.c [hunk 20]
```diff
@@ -0,0 +1,546 @@
...
+
+#include <zephyr/kernel.h>
+#include <shell/shell.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+
+#include "drv_npm6001.h"
+
+/**@brief List of register names (contents of nPM6001 NRF_DIGITAL_Type struct).
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/sample_shell_cmds.c:15]: Fixed

File: samples/pmic/npm6001/src/npm6001/drv_npm6001.h [hunk 21]
```diff
@@ -0,0 +1,449 @@
+/*
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/npm6001/drv_npm6001.h:1]: Done

File: samples/pmic/npm6001/src/sample_shell_cmds.c [hunk 22]
```diff
@@ -0,0 +1,546 @@
...
+		err = drv_npm6001_vreg_dcdc_mode_set(regulator, mode);
+		if (err) {
+			shell_warn(sh, "Failed to set mode");
+			return 0;
+		}
+
+		shell_print(sh, "Successfully set %s mode", argv[0]);
+	}
+
+	if (set_voltage && regulator == DRV_NPM6001_DCDC3) {
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/sample_shell_cmds.c:321]: Fixed

File: samples/pmic/npm6001/src/sample_shell_cmds.c [hunk 23]
```diff
@@ -0,0 +1,546 @@
...
+}
+
+static int cmd_reg_named(const struct shell *sh, size_t argc, char **argv)
+{
+	const struct reg_name_addr_pair *reg = NULL;
+	uint8_t reg_val;
+
+	if (argc == 0) {
+		shell_error(sh, "Invalid argument count");
+		return -ENOEXEC;
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/src/sample_shell_cmds.c:126]: I've observed that other shell commands use this, without questioning the usage

File: samples/pmic/npm6001/src/sample_shell_cmds.c (hunk 1 above)
Comment (nordic-auko) [reply samples/pmic/npm6001/src/sample_shell_cmds.c:534]: Fixed

**Comment by carlescufi** [comment]
@gmarull @anangl please review.

File: samples/pmic/npm6001/boards/nrf52840dk_nrf52840.overlay [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+		compatible = "gpio-keys";
+		n_int_key0: n_int_key_0 {
+			gpios = <&arduino_header 6 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;	/* D0 */
+			label = "N_INT";
+		};
+	};
+
+	aliases {
+		nint = &n_int_key0;
+	};
```
Comment (gmarull) [inline samples/pmic/npm6001/boards/nrf52840dk_nrf52840.overlay:43]: why do you need such alias?

File: samples/pmic/npm6001/boards/nrf52dk_nrf52832.overlay [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+			bias-pull-up;
+		};
+	};
+
+	i2c0_sleep_alt_alt: i2c0_sleep_alt_alt {
+		group1 {
+			psels = <NRF_PSEL(TWIM_SDA, 0, 26)>,
+				<NRF_PSEL(TWIM_SCL, 0, 27)>;
+			nordic,drive-mode = <NRF_DRIVE_S0D1>;
+			bias-pull-up;
```
Comment (gmarull) [inline samples/pmic/npm6001/boards/nrf52dk_nrf52832.overlay:26]: why is this needed in LP mode?

File: samples/pmic/npm6001/boards/nrf52dk_nrf52832.overlay [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+		n_int_key0: n_int_key_0 {
+			gpios = <&arduino_header 6 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;	/* D0 */
+			label = "N_INT";
+		};
+	};
+
+	aliases {
+		nint = &n_int_key0;
+	};
+};
```
Comment (gmarull) [inline samples/pmic/npm6001/boards/nrf52dk_nrf52832.overlay:44]: this is being replicated, why not a common file included by each board overlay?

File: samples/pmic/npm6001/prj.conf [hunk 27]
```diff
@@ -0,0 +1,27 @@
...
+CONFIG_I2C=y
+
+# Enable nPM6001 library
+CONFIG_LIB_NPM6001=y
+# Needed for certain nPM6001-EK revisions
+CONFIG_LIB_NPM6001_C3_WORKAROUND=y
+
+# Enable logging
+CONFIG_LOG=y
+CONFIG_LOG_BACKEND_RTT=n
```
Comment (gmarull) [inline samples/pmic/npm6001/prj.conf:17]:
```suggestion
```

File: samples/pmic/npm6001/prj.conf [hunk 28]
```diff
@@ -0,0 +1,27 @@
...
+
+# Enable Shell
+CONFIG_SHELL=y
+
+# Ease debugging
+CONFIG_DEBUG_OPTIMIZATIONS=y
```
Comment (gmarull) [inline samples/pmic/npm6001/prj.conf:23]:
```suggestion
```

this should not be defaut, you can provide an extra `debug.conf` with debug settings if needed

File: samples/pmic/npm6001/src/main.c [hunk 29]
```diff
@@ -0,0 +1,530 @@
...
+
+#define BUCK0_DEFAULT_VOLTAGE 3300
+#define BUCK1_DEFAULT_VOLTAGE 1200
+#define BUCK2_DEFAULT_VOLTAGE 1400
+#define BUCK3_DEFAULT_VOLTAGE 500
+#define LDO0_DEFAULT_VOLTAGE 2700
+
+K_SEM_DEFINE(n_int_sem, 0, 1);
+
+static const struct device *i2c_dev = DEVICE_DT_GET_ONE(nordic_nrf_twim);
```
Comment (gmarull) [inline samples/pmic/npm6001/src/main.c:31]: this is fragile: you need to capture to which I2C you're connecting the PMIC in DT. PMIC needs to be described in Devicetree somehow.

File: samples/pmic/npm6001/src/main.c [hunk 30]
```diff
@@ -0,0 +1,530 @@
...
+
+static uint16_t buck3_voltage;
+static bool buck3_enabled;
+static bool npm6001_gpio_set;
+
+static int lib_npm6001_platform_init(void)
+{
+	if (i2c_dev == NULL) {
+		return -ENODEV;
+	}
```
Comment (gmarull) [inline samples/pmic/npm6001/src/main.c:44]: this never holds true.

File: samples/pmic/npm6001/src/main.c [hunk 31]
```diff
@@ -0,0 +1,530 @@
...
+}
+
+static int lib_npm6001_twi_read(uint8_t *buf, uint8_t len, uint8_t reg_addr)
+{
+	int err;
+
+	if (IS_ENABLED(CONFIG_TEST)) {
+		memset(buf, 0, len);
+		return len;
+	}
```
Comment (gmarull) [inline samples/pmic/npm6001/src/main.c:56]: ?

File: samples/pmic/npm6001/src/main.c [hunk 32]
```diff
@@ -0,0 +1,530 @@
...
+
+	err = i2c_write_read(i2c_dev, LIB_NPM6001_TWI_ADDR,
+		&reg_addr, sizeof(reg_addr),
+		buf, len);
+
+	if (err == 0) {
+		/* Success: return number of bytes read */
+		return len;
+	} else {
+		return -1;
```
Comment (gmarull) [inline samples/pmic/npm6001/src/main.c:68]: ? please propagate error codes

**Review by gmarull — CHANGES_REQUESTED** [review]
a few initial comments

**Review by annwoj — APPROVED** [review]

File: samples/pmic/npm6001/boards/nrf52840dk_nrf52840.overlay [hunk 33]
```diff
@@ -0,0 +1,44 @@
...
+	pinctrl-names = "default", "sleep";
+	clock-frequency = <100000>;
+};
+
+&pinctrl {
+	i2c0_default_alt_alt: i2c0_default_alt_alt {
+		group1 {
+			psels = <NRF_PSEL(TWIM_SDA, 0, 26)>,
+				<NRF_PSEL(TWIM_SCL, 0, 27)>;
+			nordic,drive-mode = <NRF_DRIVE_S0D1>;
```
Comment (anangl) [inline samples/pmic/npm6001/boards/nrf52840dk_nrf52840.overlay:16]: This is no longer needed. The related problem [has been fixed](https://github.com/nrfconnect/sdk-zephyr/commit/0179c5349b12704b973bbd6e830ac349dda70fc5).

File: samples/pmic/npm6001/boards/nrf52840dk_nrf52840.overlay [hunk 34]
```diff
@@ -0,0 +1,44 @@
...
+	status = "okay";
+
+	pinctrl-0 = <&i2c0_default_alt_alt>;
+	pinctrl-1 = <&i2c0_sleep_alt_alt>;
+	pinctrl-names = "default", "sleep";
+	clock-frequency = <100000>;
+};
+
+&pinctrl {
+	i2c0_default_alt_alt: i2c0_default_alt_alt {
```
Comment (anangl) [inline samples/pmic/npm6001/boards/nrf52840dk_nrf52840.overlay:12]: Why `_alt_alt`?

File: samples/pmic/npm6001/src/main.c [hunk 35]
```diff
@@ -0,0 +1,530 @@
...
+#define LED_BUCK3_ON DK_LED2
+#define LED_GPIO_ON DK_LED4
+
+#define BUCK0_DEFAULT_VOLTAGE 3300
+#define BUCK1_DEFAULT_VOLTAGE 1200
+#define BUCK2_DEFAULT_VOLTAGE 1400
+#define BUCK3_DEFAULT_VOLTAGE 500
+#define LDO0_DEFAULT_VOLTAGE 2700
+
+K_SEM_DEFINE(n_int_sem, 0, 1);
```
Comment (anangl) [inline samples/pmic/npm6001/src/main.c:29]:
```suggestion
static K_SEM_DEFINE(n_int_sem, 0, 1);
```

File: samples/pmic/npm6001/src/main.c [hunk 36]
```diff
@@ -0,0 +1,530 @@
...
+
+	return 0;
+}
+
+/**
+ * @brief Function for initializing nPM6001 specific GPIO.
+ */
+static int configure_npm6001_gpio(void)
+{
+	gpio_flags_t flags = n_int.dt_flags & GPIO_ACTIVE_LOW ? GPIO_PULL_UP : GPIO_PULL_DOWN;
```
Comment (anangl) [inline samples/pmic/npm6001/src/main.c:261]:
```suggestion
```
Pull-up or pull-down should be configured in devicetree.

File: samples/pmic/npm6001/boards/nrf52840dk_nrf52840.overlay [hunk 37]
```diff
@@ -0,0 +1,44 @@
...
+};
+
+/ {
+	npm6001_irq {
+		compatible = "gpio-keys";
+		n_int_key0: n_int_key_0 {
+			gpios = <&arduino_header 6 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;	/* D0 */
+			label = "N_INT";
+		};
+	};
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/boards/nrf52840dk_nrf52840.overlay:39]: I wanted a named gpio for interrupt detection. There might be better ways of achieving this that I'm not aware of? I was initially looking for a handle to the arduino "D0" pin on the nRF5x DK's, but could not find an alias for this directly. (The D0 pin mapping is not the same for all nRF5x DKs)

File: samples/pmic/npm6001/boards/nrf52dk_nrf52832.overlay [hunk 38]
```diff
@@ -0,0 +1,44 @@
...
+
+&pinctrl {
+	i2c0_default_alt_alt: i2c0_default_alt_alt {
+		group1 {
+			psels = <NRF_PSEL(TWIM_SDA, 0, 26)>,
+				<NRF_PSEL(TWIM_SCL, 0, 27)>;
+			nordic,drive-mode = <NRF_DRIVE_S0D1>;
+			bias-pull-up;
+		};
+	};
```
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/boards/nrf52dk_nrf52832.overlay:19]: The hardware was designed a while ago, so there is not much choice with regards to the pull-ups. The internal pull-ups are 13 kOhm, which should work fine for 100 kHz clock speed with reasonably short wire lengths.

File: samples/pmic/npm6001/src/main.c (hunk 30 above)
Comment (nordic-auko) [reply samples/pmic/npm6001/src/main.c:44]: Will use `device_is_ready()` instead

File: samples/pmic/npm6001/boards/nrf52840dk_nrf52840.overlay (hunk 24 above)
Comment (nordic-auko) [reply samples/pmic/npm6001/boards/nrf52840dk_nrf52840.overlay:43]: Hm, yeah, I guess that's not strictly needed. I copied it from some reference without giving it much thought.

File: samples/pmic/npm6001/boards/nrf52dk_nrf52832.overlay (hunk 38 above)
[the remark this answers is not in the document]
Comment (nordic-auko) [reply samples/pmic/npm6001/boards/nrf52dk_nrf52832.overlay:19]: > keep things simple: just extend &i2c0_default with drive mode and pull-up.
I don't fully understand the syntax for this. Do you know of an example I can look at?

File: samples/pmic/npm6001/boards/nrf52840dk_nrf52840.overlay (hunk 34 above)
Comment (nordic-auko) [reply samples/pmic/npm6001/boards/nrf52840dk_nrf52840.overlay:12]: Renaming to `_alt`

File: samples/pmic/npm6001/src/main.c (hunk 32 above)
Comment (nordic-auko) [reply samples/pmic/npm6001/src/main.c]: Done

File: samples/pmic/npm6001/boards/nrf52dk_nrf52832.overlay (hunk 25 above)
Comment (nordic-auko) [reply samples/pmic/npm6001/boards/nrf52dk_nrf52832.overlay]: The drive-mode property is removed now. Still want to retain pull ups in all states

File: samples/pmic/npm6001/boards/nrf52dk_nrf52832.overlay (hunk 26 above)
Comment (nordic-auko) [reply samples/pmic/npm6001/boards/nrf52dk_nrf52832.overlay]: For a customer that wants to use npm6001 in a different project, I think it is easier to have a single overlay file that can be copied.

File: samples/pmic/npm6001/src/main.c (hunk 31 above)
Comment (nordic-auko) [reply samples/pmic/npm6001/src/main.c]: I wanted a simple I2C emulation without creating an emulated driver and compatible board for the zephyr i2c emulator (https://docs.zephyrproject.org/3.1.0/hardware/emulator/index.html)

**Review by carlescufi — CHANGES_REQUESTED** [review]
Blocking the introduction of a new NCS module until internal discussions take place.
